### PR TITLE
Set alignment=1 for ELF debug sections

### LINKED ISSUE #103862: Alignment of `debug_*` sections in ELF
With LinkerFlavor=gold, I am seeing this warning:
```sh
    /usr/bin/ld.gold : warning : obj/Release/net9.0/linux-musl-arm64/native/console.o: section .debug_str contains incorrectly aligned strings; the alignment of those strings won't be preserved
```
`readelf` shows we have Al=0:

```sh
$ readelf -W --headers obj/Release/net9.0/linux-musl-arm64/native/console.o | grep -E 'Name| .debug_'
  [Nr] Name              Type            Address          Off    Size   ES Flg Lk Inf Al
  [20] .debug_info       PROGBITS        0000000000000000 1274f8 03d781 00      0   0  0
  [22] .debug_str        PROGBITS        0000000000000000 1db3d1 05cbf4 00  MS  0   0  0
  [23] .debug_abbrev     PROGBITS        0000000000000000 237fc5 00019b 00      0   0  0
  [24] .debug_loc        PROGBITS        0000000000000000 238160 03dfa9 00      0   0  0
  [26] .debug_ranges     PROGBITS        0000000000000000 2f0fb9 0000a0 00      0   0  0
  [28] .debug_line       PROGBITS        0000000000000000 2f1209 00ae77 00      0   0  0
  [30] .debug_aranges    PROGBITS        0000000000000000 2fc0b0 0000b0 00      0   0 16
```
comparing with platform toolchain:
```sh
$ echo "int main(void) {}" | cc -xc -
$ readelf -W --headers a.out | grep -E 'Name| .debug_'
  [Nr] Name              Type            Address          Off    Size   ES Flg Lk Inf Al
  [20] .debug_aranges    PROGBITS        0000000000000000 010040 0000b0 00      0   0 16
  [21] .debug_info       PROGBITS        0000000000000000 0100f0 0001a0 00      0   0  1
  [22] .debug_abbrev     PROGBITS        0000000000000000 010290 00010e 00      0   0  1
  [23] .debug_line       PROGBITS        0000000000000000 01039e 00011f 00      0   0  1
  [24] .debug_frame      PROGBITS        0000000000000000 0104c0 000028 00      0   0  8
  [25] .debug_str        PROGBITS        0000000000000000 0104e8 0001dc 01  MS  0   0  1
  [26] .debug_line_str   PROGBITS        0000000000000000 0106c4 000055 01  MS  0   0  1
  [27] .debug_loclists   PROGBITS        0000000000000000 010719 000070 00      0   0  1
  [28] .debug_rnglists   PROGBITS        0000000000000000 010789 000059 00      0   0  1
```

should we set them to 1?

@filipnavara, I tested with this script which patches the alignment. That seems to make ld.gold happy:

```sh
#!/bin/bash

# List of debug sections to adjust
sections=(
    .debug_str
    .debug_info
    .debug_abbrev
    .debug_loc
    .debug_ranges
    .debug_line
    .debug_aranges
)

# Object file path
objfile="obj/Release/net9.0/linux-musl-arm64/native/console.o"

# Iterate over each section and adjust alignment
for section in "${sections[@]}"; do
    objcopy --set-section-alignment $section=1 $objfile
done

echo "Alignment adjustment completed."
```

```sh
$ cc "obj/Release/net9.0/linux-musl-arm64/native/console.o" -o "bin/Release/net9.0/linux-musl-arm64/native/console" -Wl,--version-script=obj/Release/net9.0/linux-musl-arm64/native/console.exports -Wl,--export-dynamic  -gz=zlib -fuse-ld=gold /root/.nuget/packages/runtime.linux-musl-arm64.microsoft.dotnet.ilcompiler/9.0.0-preview.6.24319.11/sdk/libbootstrapper.o /root/.nuget/packages/runtime.linux-musl-arm64.microsoft.dotnet.ilcompiler/9.0.0-preview.6.24319.11/sdk/libRuntime.WorkstationGC.a /root/.nuget/packages/runtime.linux-musl-arm64.microsoft.dotnet.ilcompiler/9.0.0-preview.6.24319.11/sdk/libeventpipe-disabled.a /root/.nuget/packages/runtime.linux-musl-arm64.microsoft.dotnet.ilcompiler/9.0.0-preview.6.24319.11/sdk/libstandalonegc-disabled.a /root/.nuget/packages/runtime.linux-musl-arm64.microsoft.dotnet.ilcompiler/9.0.0-preview.6.24319.11/sdk/libstdc++compat.a /root/.nuget/packages/runtime.linux-musl-arm64.microsoft.dotnet.ilcompiler/9.0.0-preview.6.24319.11/framework/libSystem.Native.a /root/.nuget/packages/runtime.linux-musl-arm64.microsoft.dotnet.ilcompiler/9.0.0-preview.6.24319.11/framework/libSystem.IO.Compression.Native.a /root/.nuget/packages/runtime.linux-musl-arm64.microsoft.dotnet.ilcompiler/9.0.0-preview.6.24319.11/framework/libSystem.Net.Security.Native.a /root/.nuget/packages/runtime.linux-musl-arm64.microsoft.dotnet.ilcompiler/9.0.0-preview.6.24319.11/framework/libSystem.Security.Cryptography.Native.OpenSsl.a -g -Wl,-rpath,'$ORIGIN' -Wl,--build-id=sha1 -Wl,--as-needed -pthread -ldl -lz -lrt -lm -pie -Wl,-pie -Wl,-z,relro -Wl,-z,now -Wl,-z,stack-size=1572864 -Wl,--eh-frame-hdr -Wl,--discard-all

# no warning
```

### PR DESCRIPTION
**Pull request by am11**
Fix #103862

### CODE REVIEW

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ObjectWriter/ElfObjectWriter.cs [hunk 1]
```diff
@@ -141,6 +141,7 @@ private protected override void CreateSection(ObjectNodeSection section, string
                 {
                     Type = type,
                     Flags = flags,
+                    Alignment = section.Type == SectionType.Debug ? 1 : 0
```
Comment (filipnavara) [inline src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ObjectWriter/ElfObjectWriter.cs:144]:
```suggestion
                    Alignment = 1
```

I think we can skip this check. Presumably any section will get some alignment set through `UpdateSectionAlignment` based on the alignment requirement of the data inside.

**Review by filipnavara — APPROVED** [review]
LGTM if the tests pass

**Comment by agocke** [comment]
We have a smoke test on dwarfdump. Could we also add some validation for this in the same place?

**Comment by jkotas** [comment]
> We have a smoke test on dwarfdump. Could we also add some validation for this in the same place?

dwarfdump does not complain about this detail. We would have to dump the dwarf file, parse the dump and verify that the alignment of debug sections is non-zero. It sounds too complicated to be worth it.

It would be better to add validation that linking with `LinkerFlavor=gold` produces working binary without any warnings. It would catch this and all other potential gold linker specific issues.

**Comment by am11** [comment]
I agree that a basic app validation would be nice with all four linkers. Since we don't have gold linker installed in the CI, it can be added later. Besides section headers are rarely changed.

**Review by MichalStrehovsky — APPROVED** [review]
Testing with all sorts of linkers is a known gap unfortunately and this would be covered if we start doing that.

**Pull request merged**